packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js

```diff
@@ -0,0 +1,189 @@
+/**
+ * External dependencies
+ */
+import { difference, isEqual, without } from 'lodash';
+
+const { wp } = window;
+
+function parseWidgetId( widgetId ) {
+	const matches = widgetId.match( /^(.+)-(\d+)$/ );
+	if ( matches ) {
+		return {
+			idBase: matches[ 1 ],
+			number: parseInt( matches[ 2 ], 10 ),
+		};
+	}
+
+	// Likely an old single widget.
+	return { idBase: widgetId };
+}
+
+function widgetIdToSettingId( widgetId ) {
+	const { idBase, number } = parseWidgetId( widgetId );
+	if ( number ) {
+		return `widget_${ idBase }[${ number }]`;
+	}
+
+	return `widget_${ idBase }`;
+}
+
+function parseSettingId( settingId ) {
+	const matches = settingId.match( /^widget_(.+)(?:\[(\d+)\])$/ );
+	if ( matches ) {
+		return {
+			idBase: matches[ 1 ],
+			number: parseInt( matches[ 2 ], 10 ),
+		};
+	}
+
+	return { idBase: settingId };
+}
+
+function settingIdToWidgetId( settingId ) {
+	const { idBase, number } = parseSettingId( settingId );
+	if ( number ) {
+		return `${ idBase }-${ number }`;
+	}
+
+	return idBase;
+}
+
+export default class SidebarAdapter {
+	constructor( setting, allSettings ) {
+		this.setting = setting;
+		this.allSettings = allSettings;
+
+		this.subscribers = new Set();
+
+		this._handleSettingChange = this._handleSettingChange.bind( this );
```

Comment:
Maybe we should bite the bullet and add an observable library lol. It's come up a few times in some PRs I've been a part of. Not right now though.

Reply:
I think maybe we can further simplify this a lot to the point that we don't need to introduce a observable library though 🤔 . Let's see how it goes 😆 .